### PR TITLE
Add gradient clipping to Static Capture Utilities

### PR DESCRIPTION
**Pull request by dallasfoster**
# Modulus Pull Request

## Description
Adds `gradient_clip_norm` to the static capture utilities. This should be backward compatible with existing examples.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/utils/capture.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ def __init__(
         use_gradscaler: bool = True,
         cuda_graph_warmup: int = 11,
         amp_type: Union[float16, bfloat16] = torch.float16,
+        gradient_clip_norm: float | None = None,
```
Comment (loliverhennigh) [inline modulus/utils/capture.py:66]: I did not know you could do `float | None` instead of `Union[float, None]`. Crazy...

File: test/utils/test_capture.py [hunk 2]
```diff
@@ -64,6 +64,7 @@ def logger():
     "use_amp, amp_type",
     [(True, torch.float16), (True, torch.bfloat16), (False, torch.float16)],
 )
+@pytest.mark.parametrize("gradient_clip_norm", [None, 4.0])
```
Comment (loliverhennigh) [inline test/utils/test_capture.py:67]: Might be good to have the unit test somehow actually check if the gradients got clipped. Not super needed though.

**Review by loliverhennigh — APPROVED** [review]
Looks good to me. Small commit on the unit test but not super needed.

File: modulus/utils/capture.py [hunk 3]
```diff
@@ -251,6 +253,13 @@ def _amp_forward(self, *args, **kwargs) -> Any:
         if not self.eval:
             # In training mode output should be the loss
             self.scaler.scale(output).backward()
+            if self.gradient_clip_norm is not None:
+                if self.use_autocast:
```
Comment (Alexey-Kamenev) [inline modulus/utils/capture.py:257]: I think this check is redundant - `GradScaler.unscale_` will check whether the scaler is enabled. Ok to keep it anyway for consistency with the other code.

File: modulus/utils/capture.py [hunk 4]
```diff
@@ -348,6 +357,8 @@ class StaticCaptureTraining(_StaticCapture):
         Number of warmup steps for cuda graphs, by default 11
     amp_type : Union[float16, bfloat16], optional
         Auto casting type for AMP, by default torch.float16
+    gradient_clip_norm : Union[float, None], optional
+        Threshold for gradient clipping
     label : Optional[str, None], optional
```
Comment (Alexey-Kamenev) [inline modulus/utils/capture.py:362]: Nit: probably meant `Optional[str]`, no?

File: modulus/utils/capture.py [hunk 5]
```diff
@@ -348,6 +357,8 @@ class StaticCaptureTraining(_StaticCapture):
         Number of warmup steps for cuda graphs, by default 11
     amp_type : Union[float16, bfloat16], optional
         Auto casting type for AMP, by default torch.float16
+    gradient_clip_norm : Union[float, None], optional
```
Comment (Alexey-Kamenev) [inline modulus/utils/capture.py:360]: Nit: use either `Optional[float]` or `float | None`?

**Review by Alexey-Kamenev — APPROVED** [review]
LGTM!

File: modulus/utils/capture.py (hunk 1 above)
Comment (dallasfoster) [reply modulus/utils/capture.py]: Opted to use `Optional[float]` instead.

File: test/utils/test_capture.py (hunk 2 above)
Comment (dallasfoster) [reply test/utils/test_capture.py]: Done.

File: modulus/utils/capture.py (hunk 3 above)
Comment (dallasfoster) [reply modulus/utils/capture.py]: I will remove it, thanks.

File: modulus/utils/capture.py (hunk 4 above)
Comment (dallasfoster) [reply modulus/utils/capture.py]: Good point, I standardize the rest of the docstrings to match.

File: modulus/utils/capture.py (hunk 5 above)
Comment (dallasfoster) [reply modulus/utils/capture.py]: Good point, I will standard to using `Optional` instead.
